# Preprocessing

First, let's get all of the imports out of the way:

In [1]:
#Dataframe manipulation library
import pandas as pd
#Math functions, we'll only need the sqrt function so let's import only that
from math import sqrt
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

Now let's read each file into their Dataframes:

In [2]:
#Storing the artists information into a pandas dataframe
artists_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/artists.csv')
#Storing the tags information into a pandas dataframe
tags_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/tags.csv')
#Storing the user artists information into a pandas dataframe
user_artists_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/user_artists.csv')
#Storing the user tagged artists information into a pandas dataframe
user_taggedartists_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/user_taggedartists.csv')

In [3]:
print('Shape of artists_df:',artists_df.shape)
print('Shape of tags_df:',tags_df.shape)
print('Shape of user_artists_df:',user_artists_df.shape)
print('Shape of user_taggedartists_df:',user_taggedartists_df.shape)

Shape of artists_df: (17632, 4)
Shape of tags_df: (11946, 2)
Shape of user_artists_df: (92834, 3)
Shape of user_taggedartists_df: (186479, 6)


In [4]:
artists_df.describe(include = 'all')[['id', 'name']]

id   name
count   17632.000000  17632
unique           NaN  17632
top              NaN  Dozer
freq             NaN      1
mean     9156.636853    NaN
std      5392.515296    NaN
min         1.000000    NaN
25%      4501.750000    NaN
50%      9048.500000    NaN
75%     13769.250000    NaN
max     18745.000000    NaN

In [5]:
tags_df.describe(include = 'all')

tagID tagValue
count   11946.000000    11946
unique           NaN    11944
top              NaN   #NAME?
freq             NaN        3
mean     6242.315336      NaN
std      3667.498057      NaN
min         1.000000      NaN
25%      3036.250000      NaN
50%      6210.500000      NaN
75%      9460.750000      NaN
max     12648.000000      NaN

In [6]:
user_artists_df.describe(include = 'all')

userID      artistID        weight
count  92834.000000  92834.000000   92834.00000
mean    1037.010481   3331.123145     745.24393
std      610.870436   4383.590502    3751.32208
min        2.000000      1.000000       1.00000
25%      502.000000    436.000000     107.00000
50%     1029.000000   1246.000000     260.00000
75%     1568.000000   4350.000000     614.00000
max     2100.000000  18745.000000  352698.00000

In [7]:
user_taggedartists_df.describe(include = 'all')

userID       artistID          tagID            day  \
count  186479.000000  186479.000000  186479.000000  186479.000000   
mean     1035.600137    4375.845328    1439.582913       1.095566   
std       622.461272    4897.789595    2775.340279       0.712813   
min         2.000000       1.000000       1.000000       1.000000   
25%       488.000000     686.000000      79.000000       1.000000   
50%      1021.000000    2203.000000     195.000000       1.000000   
75%      1624.000000    6714.000000     887.000000       1.000000   
max      2100.000000   18744.000000   12647.000000       9.000000   

               month           year  
count  186479.000000  186479.000000  
mean        6.524215    2008.809791  
std         3.486855       1.410062  
min         1.000000    1956.000000  
25%         3.000000    2008.000000  
50%         7.000000    2009.000000  
75%        10.000000    2010.000000  
max        12.000000    2011.000000

### Merging columns

<p>Merge columns (artistid, artistname, tagid) and store them in a new dataframe artists.</p>

In [8]:
#rename the column id in artists_df to artistID to make easier to join the dataframes.
artists_df.rename(columns = {'id': 'artistID'}, inplace = True)

#initially merge artists_df and user_taggedartists_df as they both contain artistID
artists = pd.merge(artists_df, user_taggedartists_df, on='artistID', how='inner')

#Then merge artists and tags_df
artists = pd.merge(artists, tags_df, on = 'tagID', how = 'inner')
print(artists.shape)
artists.head()

(184941, 10)


artistID                      name  \
0         1              MALICE MIZER   
1         8               DIR EN GREY   
2       374        宇多田ヒカル   
3      1929  ASIAN KUNG-FU GENERATION   
4         1              MALICE MIZER   

                                                 url  \
0              http://www.last.fm/music/MALICE+MIZER   
1               http://www.last.fm/music/DIR+EN+GREY   
2  http://www.last.fm/music/%E5%AE%87%E5%A4%9A%E7...   
3  http://www.last.fm/music/ASIAN+KUNG-FU+GENERATION   
4              http://www.last.fm/music/MALICE+MIZER   

                                          pictureURL  userID  tagID  day  \
0    http://userserve-ak.last.fm/serve/252/10808.jpg      37    552    1   
1  http://userserve-ak.last.fm/serve/252/46968835...      37    552    1   
2  http://userserve-ak.last.fm/serve/252/55544057...      37    552    1   
3   http://userserve-ak.last.fm/serve/252/358164.jpg      37    552    1   
4    http://userserve-ak.last.fm/serve/252/10808.jpg     274   1219    1   

   month  year tagValue  
0      7  2008   weeabo  
1      7  2008   weeabo  
2      7  2008   weeabo  
3      7  2008   weeabo  
4      5  2008    jrock

Drop unnecessary columns.

In [9]:
artists.drop(['url', 'pictureURL', 'userID', 'day', 'month', 'year', 'tagValue'], inplace = True, axis = 1)
artists.head(10)

artistID                      name  tagID
0         1              MALICE MIZER    552
1         8               DIR EN GREY    552
2       374        宇多田ヒカル    552
3      1929  ASIAN KUNG-FU GENERATION    552
4         1              MALICE MIZER   1219
5         8               DIR EN GREY   1219
6         8               DIR EN GREY   1219
7         8               DIR EN GREY   1219
8       376                 UVERworld   1219
9       376                 UVERworld   1219

Since the name column contains non-alphanumeric values, initally replace those values with NaN and then drop rows that contain NaN values.

In [10]:
artists['name'].replace(r'\W', np.nan, regex=True, inplace=True)
artists.dropna(inplace = True)

In [11]:
artists.head(10)

artistID       name  tagID
8        376  UVERworld   1219
9        376  UVERworld   1219
10       390      GACKT   1219
13      1857        YUI   1219
14      1872       Flow   1219
23      4980     Olivia   1219
25      5077       hide   1219
33      8396      12012   1219
34      8397     VIDOLL   1219
35      8404      Sadie   1219

In [12]:
artists.reset_index(drop = True)

artistID          name  tagID
0           376     UVERworld   1219
1           376     UVERworld   1219
2           390         GACKT   1219
3          1857           YUI   1219
4          1872          Flow   1219
5          4980        Olivia   1219
6          5077          hide   1219
7          8396         12012   1219
8          8397        VIDOLL   1219
9          8404         Sadie   1219
10        12801   Megaromania   1219
11        12803         ViViD   1219
12        17719           Miz   1219
13          376     UVERworld    139
14          376     UVERworld    139
15          376     UVERworld    139
16          376     UVERworld    139
17          376     UVERworld    139
18          390         GACKT    139
19          390         GACKT    139
20          390         GACKT    139
21          390         GACKT    139
22          406          HYDE    139
23          406          HYDE    139
24         1857           YUI    139
25         1857           YUI    139
26         1857           YUI    139
27         1872          Flow    139
28         1872          Flow    139
29         1872          Flow    139
...         ...           ...    ...
54810     18232  THEreminGIRL  12232
54811     18229         VFSix  12233
54812     18232  THEreminGIRL  12233
54813     18229         VFSix  12234
54814     18232  THEreminGIRL  12234
54815     18229         VFSix  12235
54816     18232  THEreminGIRL  12235
54817     18229         VFSix  12236
54818     18232  THEreminGIRL  12236
54819     18283        Expose   5490
54820     18320          Lava   4747
54821     18341     Castanets   7705
54822     18424      Amycanbe   2633
54823     18424      Amycanbe  10818
54824     18424      Amycanbe  12455
54825     18424      Amycanbe  12456
54826     18424      Amycanbe  12457
54827     18424      Amycanbe  12458
54828     18424      Amycanbe  12459
54829     18424      Amycanbe  12460
54830     18424      Amycanbe  12461
54831     18615         YEFFI  12616
54832     18615         YEFFI  12617
54833     18615         YEFFI  12618
54834     18630  Doppelganger  12633
54835     18736     LOSTFREEQ   3661
54836     18736     LOSTFREEQ   3665
54837     18736     LOSTFREEQ   3666
54838     18736     LOSTFREEQ   3667
54839     18736     LOSTFREEQ   3672

[54840 rows x 3 columns]

In [13]:
artists.describe(include = 'all')

artistID     name         tagID
count   54840.000000    54840  54840.000000
unique           NaN     3407           NaN
top              NaN  Madonna           NaN
freq             NaN      730           NaN
mean     4157.318071      NaN   1345.595678
std      4756.374220      NaN   2713.337728
min         6.000000      NaN      1.000000
25%       707.000000      NaN     73.000000
50%      2094.000000      NaN    187.000000
75%      6160.000000      NaN    824.000000
max     18736.000000      NaN  12647.000000

As there are 54,840 records, which is quite a lot, let's take sample of 15% of the data for further analysis.

In [14]:
artists_sample = artists.sample(frac = 0.15)

#Let's check if artists_sample contains 15% of the artists data
if(0.15 * (len(artists)) == len(artists_sample)):
    print('artists_sample data frame contains 15% of artists data!')
    
print(artists_sample.shape)
artists_sample.describe(include = 'all')

artists_sample data frame contains 15% of artists data!
(8226, 3)


artistID     name         tagID
count    8226.000000     8226   8226.000000
unique           NaN     2147           NaN
top              NaN  Madonna           NaN
freq             NaN       99           NaN
mean     4186.444931      NaN   1346.392293
std      4775.810489      NaN   2690.923745
min         6.000000      NaN      1.000000
25%       707.000000      NaN     73.000000
50%      2101.000000      NaN    191.000000
75%      6220.250000      NaN    829.000000
max     18736.000000      NaN  12644.000000

Let's convert tagID into a string, to make it easier for creating dummy variables for the column.

In [15]:
artists_sample['tagID'] = artists_sample['tagID'].astype('str')
print(artists_sample.dtypes)

artists_sample.describe(include = 'all')

artistID     int64
name        object
tagID       object
dtype: object


artistID     name tagID
count    8226.000000     8226  8226
unique           NaN     2147  1636
top              NaN  Madonna    73
freq             NaN       99   317
mean     4186.444931      NaN   NaN
std      4775.810489      NaN   NaN
min         6.000000      NaN   NaN
25%       707.000000      NaN   NaN
50%      2101.000000      NaN   NaN
75%      6220.250000      NaN   NaN
max     18736.000000      NaN   NaN

<p>Since keeping tags in a list format isn't optimal for the content-based recommendation system technique, we will use the One Hot Encoding technique to convert the list of tag values to a vector where each column corresponds to one possible value of the feature.</p>

<p>This encoding is needed for feeding categorical data. In this case, we store every different tag value in columns that contain either 1 or 0. 1 shows that an artist has that genre and 0 shows that it doesn't.</p>

<p>Let's also store this dataframe in another variable since tags won't be important for our first recommendation system.</p>

In [16]:
#Copying the movie dataframe into a new one since we won't need to use the genre information in our first case.
artistsWithTags_df = artists_sample.copy()

artistsWithTags_df = pd.get_dummies(data = artists_sample, columns = ['tagID'])
print(artistsWithTags_df.shape)

artistsWithTags_df.head()

(8226, 1638)


artistID           name  tagID_1  tagID_10  tagID_10006  tagID_1001  \
109395      4574           Ania        0         0            0           0   
118925       644           Inna        0         0            0           0   
92279       5188  Shapeshifters        0         0            0           0   
94425       8802         Kitaro        0         0            0           0   
150139      4375           MxPx        0         0            0           0   

        tagID_1002  tagID_1003  tagID_1006  tagID_10061  ...  tagID_9954  \
109395           0           0           0            0  ...           0   
118925           0           0           0            0  ...           0   
92279            0           0           0            0  ...           0   
94425            0           0           0            0  ...           0   
150139           0           0           0            0  ...           0   

        tagID_9955  tagID_996  tagID_9960  tagID_9961  tagID_9967  tagID_9972  \
109395           0          0           0           0           0           0   
118925           0          0           0           0           0           0   
92279            0          0           0           0           0           0   
94425            0          0           0           0           0           0   
150139           0          0           0           0           0           0   

        tagID_9973  tagID_9991  tagID_9997  
109395           0           0           0  
118925           0           0           0  
92279            0           0           0  
94425            0           0           0  
150139           0           0           0  

[5 rows x 1638 columns]

In [17]:
artistsWithTags_df.reset_index(drop = True)

artistID            name  tagID_1  tagID_10  tagID_10006  tagID_1001  \
0         4574            Ania        0         0            0           0   
1          644            Inna        0         0            0           0   
2         5188   Shapeshifters        0         0            0           0   
3         8802          Kitaro        0         0            0           0   
4         4375            MxPx        0         0            0           0   
5        10071      Mariottide        0         0            0           0   
6         1110            Rush        0         0            0           0   
7         3411             M83        0         0            0           0   
8         6769          Cartel        0         0            0           0   
9          859          Pixies        0         0            0           0   
10         928       Helloween        0         0            0           0   
11        1263       Tristania        0         0            0           0   
12        1045          Fresno        0         0            0           0   
13        2289       Stereolab        0         0            0           0   
14          88        Gorillaz        0         0            0           0   
15         371         Skillet        0         0            0           0   
16           6       Moonspell        0         0            0           0   
17          67         Madonna        0         0            0           0   
18        1881             XTC        0         0            0           0   
19        3445          Danzig        0         0            0           0   
20         978       Disturbed        1         0            0           0   
21          65        Coldplay        0         0            0           0   
22        1044        Slipknot        0         0            0           0   
23        3884        Slowdive        0         0            0           0   
24         493    Lostprophets        0         0            0           0   
25         907       Timbaland        0         0            0           0   
26         959           Queen        0         0            0           0   
27        1866      Yellowcard        0         0            0           0   
28         191     OneRepublic        0         0            0           0   
29         434           Spoon        0         0            0           0   
...        ...             ...      ...       ...          ...         ...   
8196     10987          Brodka        0         0            0           0   
8197     10737       Blindside        0         0            0           0   
8198      3308      Supertramp        0         0            0           0   
8199       108    Magnitophono        0         0            0           0   
8200       824       Ensiferum        0         0            0           0   
8201      3317           Drake        0         0            0           0   
8202     16875        Bankrupt        0         0            0           0   
8203       769            Isis        0         0            0           0   
8204     10650          Oceana        0         0            0           0   
8205       755  65daysofstatic        0         0            0           0   
8206      7172            Enam        0         0            0           0   
8207      4165          Poison        0         0            0           0   
8208       511              U2        0         0            0           0   
8209      3469      Supergrass        0         0            0           0   
8210       154       Radiohead        0         0            0           0   
8211       730          Eagles        0         0            0           0   
8212       707       Metallica        0         0            0           0   
8213       506        Anathema        1         0            0           0   
8214       512        Deftones        0         0            0           0   
8215        67         Madon

Now, let's take a look at user_artists_df dataframe.

In [18]:
print(user_artists_df.shape)
user_artists_df.head()

(92834, 3)


userID  artistID  weight
0       2        51   13883
1       2        52   11690
2       2        53   11351
3       2        54   10300
4       2        55    8983

In [19]:
df = pd.merge(user_artists_df, artistsWithTags_df, on='artistID')
print(df.shape)

(490469, 1640)


In [20]:
df.head()

userID  artistID  weight       name  tagID_1  tagID_10  tagID_10006  \
0       2        52   11690  Morcheeba        0         0            0   
1       2        52   11690  Morcheeba        0         0            0   
2       2        52   11690  Morcheeba        0         0            0   
3       2        52   11690  Morcheeba        0         0            0   
4       2        52   11690  Morcheeba        0         0            0   

   tagID_1001  tagID_1002  tagID_1003  ...  tagID_9954  tagID_9955  tagID_996  \
0           0           0           0  ...           0           0          0   
1           0           0           0  ...           0           0          0   
2           0           0           0  ...           0           0          0   
3           0           0           0  ...           0           0          0   
4           0           0           0  ...           0           0          0   

   tagID_9960  tagID_9961  tagID_9967  tagID_9972  tagID_9973  tagID_9991  \
0           0           0           0           0           0           0   
1           0           0           0           0           0           0   
2           0           0           0           0           0           0   
3           0           0           0           0           0           0   
4           0           0           0           0           0           0   

   tagID_9997  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 1640 columns]

In [21]:
## Checking a sample of how many times users listened to artists
print("Number of users who provided ratings:", len(df['userID'].unique()))
print(df.groupby(["userID"])["weight"].agg([max,min,np.mean,np.median,len])[0:20])

Number of users who provided ratings: 1874
          max   min          mean  median  len
userID                                        
2       11690  1315   3688.060345  3301.0  464
3       13176    67   3591.814815   269.0   27
4         826   202    375.717996   292.0  539
5         884   129    444.956790   265.0  486
6          27    13     16.446809    14.0   47
7       43864   593  14765.765873  2540.0  252
8        1039   178    531.902597   428.0  154
9        1113   249    622.971154   590.0  104
10       1297   157    424.502994   177.0  167
11       1318   229    475.936242   370.0  298
12      31617   927   5998.129973  3072.0  377
13          9     2      5.539326     6.0  356
15        751   242    599.359606   740.0  203
16        487   141    228.931818   154.0   44
17       1544   115    537.336478   303.0  318
18        838   204    437.546778   496.0  481
20        375    98    182.062500   154.0  128
21        732    98    378.906977   369.0  258
22       2501   5

In [22]:
import sklearn.metrics.pairwise as pw
from scipy import sparse
from sklearn.metrics.pairwise import pairwise_distances

def item_based_recom(input_dataframe,input_artist_name):    
    pivot_item_based = pd.pivot_table(input_dataframe,
                                      index='name',
                                      columns=['userID'], values='weight')  
    sparse_pivot = sparse.csr_matrix(pivot_item_based.fillna(0))
    recommender = pw.cosine_similarity(sparse_pivot)
    recommender_df = pd.DataFrame(recommender, 
                                  columns=pivot_item_based.index,
                                  index=pivot_item_based.index)
    ## Item Rating Based Cosine Similarity
    cosine_df = pd.DataFrame(recommender_df[artist_name].sort_values(ascending=False))
    cosine_df.reset_index(level=0, inplace=True)
    cosine_df.columns = ['Artist','cosine_sim']
    return cosine_df

In [26]:
artist_name ='Madonna'

In [27]:
def generate_recommendations(df,artist_name,top_results=5):
    print("Artists you might enjoy based on artist you listened", artist_name)
    ## Item Rating Based Cosine Similarity
    cos_sim = item_based_recom(df,artist_name)
    display(cos_sim[1:top_results+1])

In [28]:
generate_recommendations(df,artist_name)

Artists you might enjoy based on artist you listened Madonna


Artist  cosine_sim
1         Cher    0.409757
2  Tribalistas    0.336903
3         Knob    0.333841
4    Anastacia    0.271026
5        Flora    0.255404